functorch/_src/aot_autograd.py

```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
 #
 # TODO: Provide a faster version of this that assumes flat arguments
 # (so no pytree necessary)
-def detach_and_functionalize_pure(f, preserve_requires_grad=True):
+def run_functionalized_fw_and_collect_metadata(f):
+    def to_fun(t):
+        if isinstance(t, Tensor):
+            return torch._to_functional_tensor(t, mirror_autograd_meta=True)
+        else:
+            return t
+
+    def from_fun(t):
+        if not isinstance(t, Tensor) or not torch._is_functional_tensor(t):
+            return t
+        torch._sync(t)
+        return torch._from_functional_tensor(t)
+
     @wraps(f)
-    def inner(*args, **kwargs):
-        def to_fun(t):
-            if isinstance(t, Tensor):
-                r = torch._to_functional_tensor(t)
-                # NB: r is a leaf; it has no grad_fn relating
-                # it to t.  If t has autograd metadata, that
-                # metadata was preserved *inside* the r wrapper
-                if preserve_requires_grad:
-                    r.requires_grad = t.requires_grad
-                return r
-            else:
-                return t
+    def inner(*args):
+        # This function is meant to be run with the forward, which expects a flat list of tensor/symint/other args.
+        assert all(isinstance(a, torch.Tensor) or type(a) in KNOWN_TYPES for a in args)
```

Comment:
Are you sure you didn't want an `isinstance(a, KNOWN_TYPES)` test, where `KNOWN_TYPES` must be a tuple